File: Headless/Headless/Wall.cs
using nkast.Aether.Physics2D.Common;
using System;
using System.Collections.Generic;
using System.Text;

namespace Headless
{
    internal class Wall
    {
        public readonly Vector2 leftPoint;
        public readonly Vector2 rightPoint;
        public readonly Vector2 origin;

        public readonly Vector2 orthogonalVector;
        public readonly Vector2 reverseOrthogonalVector;
        public readonly Vector2 inlineVector;

        public readonly float length;

        public Wall(Vector2 leftPoint, Vector2 rightPoint)
        {
            this.leftPoint = leftPoint;
            this.rightPoint = rightPoint;

            this.origin = (leftPoint + rightPoint) / 2;

            this.inlineVector = Helpers.normalize(leftPoint - rightPoint);
            this.orthogonalVector = new Vector2(-inlineVector.Y, inlineVector.X);
            this.reverseOrthogonalVector = -orthogonalVector;

            this.length = Helpers.norm(rightPoint - leftPoint);
        }
        public Wall(Vector2 origin, float length, float angleDegrees)
        {
            this.origin = origin;
            var leftNorm = new Vector2(Mathf.Cos(angleDegrees * Helpers.deg2rad), Mathf.Sin(angleDegrees * Helpers.deg2rad));
            var rightNorm = -leftNorm;
            this.leftPoint = origin + leftNorm * length / 2;
            this.rightPoint = origin + rightNorm * length / 2;

            this.inlineVector = Helpers.normalize(leftPoint - rightPoint);
            this.orthogonalVector = new Vector2(-inlineVector.Y, inlineVector.X);
            this.reverseOrthogonalVector = -orthogonalVector;

            this.length = length;
        }

        //CONSTANTS
        public static Wall floor = new Wall(new Vector2(-200, 0), new Vector2(200, 0));
        public static Vector2 hubOrigin = new Vector2(0, 1.8288f);
        public static Wall[] hub = CreateRectangle(1.688288f, 1.383801f, new Vector2(0, -1.1368f) + hubOrigin);
        public static Wall hubTop = new Wall(new Vector2(1.06f/2, 0) + hubOrigin, new Vector2(-1.06f, 0) + hubOrigin);
        public static Wall hubBottom = new Wall(new Vector2(0.605f/2, -0.39f) + hubOrigin, new Vector2(-0.605f, -0.39f) + hubOrigin);
        public static Wall hub1 = new Wall(new Vector2(0.415f, -0.2f) + hubOrigin, 0.47f, -30f);
        public static Wall hub2 = new Wall(new Vector2(-0.415f, -0.2f) + hubOrigin, 0.47f, 30f);

        public static Wall[] allKillWalls = new Wall[]
        {
            floor,
            hub[0], hub[1], hub[2], hub[3],
            hub1, hub2
        };
        
        public static Wall[] CreateRectangle(float width, float height, Vector2 center)
        {
            var walls = new Wall[4];
            walls[0] = new Wall(center + new Vector2(-width / 2, +height / 2), center + new Vector2(-width / 2, -height / 2));
            walls[1] = new Wall(center + new Vector2(+width / 2, -height / 2), center + new Vector2(+width / 2, +height / 2));
            walls[2] = new Wall(center + new Vector2(+width / 2, +height / 2), center + new Vector2(-width / 2, +height / 2));
            walls[3] = new Wall(center + new Vector2(-width / 2, -height / 2), center + new Vector2(+width / 2, -height / 2));
            return walls;
        }
    }
}

File: CLI/Rust/src/wall.rs
use crate::vec2::Vec2;
use lazy_static::lazy_static;

lazy_static! {
    pub static ref FLOOR: Wall = Wall::new(Vec2::new(200.0, 0.0), Vec2::new(-200.0, 0.0));
    pub static ref HUB_ORIGIN: Vec2 = Vec2::new(0.0, 1.8288);
    pub static ref HUB: Vec<Wall> = Wall::create_rect(1.688288, 1.383801, Vec2::new(0.0, -1.1368) + *HUB_ORIGIN);
    pub static ref HUB_TOP: Wall = Wall::new(Vec2::new(-1.06/2.0, 0.0) + *HUB_ORIGIN, Vec2::new(1.06/2.0, 0.0) + *HUB_ORIGIN);
    pub static ref HUB_BOTTOM: Wall = Wall::new(Vec2::new(-0.605/2.0, -0.39) + *HUB_ORIGIN, Vec2::new(0.605/2.0, -0.39) + *HUB_ORIGIN);
    pub static ref HUB_SIDE_INNER: Wall = Wall::from_origin(Vec2::new(-0.415, -0.2) + *HUB_ORIGIN, 0.47, 30.0);
    pub static ref HUB_SIDE_OUTER: Wall = Wall::from_origin(Vec2::new(0.415, -0.2) + *HUB_ORIGIN, 0.47, -30.0);

    pub static ref ALL_KILL_WALLS: Vec<Wall> = vec![
        *FLOOR,
        HUB[0], HUB[1], HUB[2], HUB[3],
        *HUB_SIDE_INNER,
    ];

    pub static ref HUB_SIDES: Vec<Wall> = vec![
        *HUB_SIDE_OUTER
    ];
}

#[derive(Copy, Clone)]
pub struct Wall {
    pub left_point: Vec2,
    pub right_point: Vec2,
    pub origin: Vec2,
    pub orthogonal: Vec2,
    pub reverse_orthogonal: Vec2,
    pub inline: Vec2,
    pub length: f64
}

impl Wall {
    pub fn new(left_point: Vec2, right_point: Vec2) -> Self {
        let inline = (left_point - right_point).normalize();
        let orthogonal = Vec2::new(-inline.y, inline.x);
        Self {
            left_point,
            right_point,
            origin: (left_point + right_point) / 2.0,
            orthogonal,
            reverse_orthogonal: -orthogonal,
            inline,
            length: (right_point - left_point).norm()
        }
    }

    pub fn from_origin(origin: Vec2, length: f64, angle_degrees: f64) -> Self {
        let left_unit = Vec2::new(angle_degrees.to_radians().cos(), angle_degrees.to_radians().sin());
        let right_unit = -left_unit;

        let left_point = origin + left_unit * length / 2.0;
        let right_point = origin + right_unit * length / 2.0;

        let inline = (left_point - right_point).normalize();
        let orthogonal = Vec2::new(-inline.y, inline.x);

        Self {
            left_point,
            right_point,
            origin,
            orthogonal,
            reverse_orthogonal: -orthogonal,
            inline,
            length
        }
    }

    pub fn create_rect(width: f64, height: f64, center: Vec2) -> Vec<Self> {
        let side1 = Self::new(center + Vec2::new(-width / 2.0, height / 2.0), center + Vec2::new(-width / 2.0, -height / 2.0));
        let side2 = Self::new(center + Vec2::new(width / 2.0, -height / 2.0), center + Vec2::new(width / 2.0, height / 2.0));
        let side3 = Self::new(center + Vec2::new(width / 2.0, height / 2.0), center + Vec2::new(-width / 2.0, height / 2.0));
        let side4 = Self::new(center + Vec2::new(-width / 2.0, -height / 2.0), center + Vec2::new(width / 2.0, -height / 2.0));
        return vec![side1, side2, side3, side4];
    }
}